### PR TITLE
[Do not merge] Add review stacks

### PR DESCRIPTION
**Pull request by indiebrain**
Tracks the proposed upstream changes to "Add Review Stacks." This PR IS NOT intended to be merged into `powerhome/master`, rather these changes should come into `powerhome/master` via a pull / merge from `shopify/master` if/when https://github.com/Shopify/shipit-engine/pull/1102 is accepted.

Add Review Stacks
-----------------

This change set represents a months long effort to build, utilize, evolve, and now - hopefully - contribute-for-general-use the concept of "Review Stacks" to shipit-engine.

As discussed in issue #960, it is our desire to use shipit-engine to provide something akin to Heroku Pipeline Review Applications - a dynamically managed application instance. Review Stacks are intended to run the code of a Github Pull Request as a disposable application instance. 

Terminology
-----------

- Review Environment, a dynamically (de)provisioned set of resources used to host an instance of an application
- Review Stack, a dynamically managed stack which relates a Github Pull Request to a Review Environment

Details
-------

Detailed feature documentation, and user guide, are provided in `docs/review_stacks.md`.

The Review Stack feature's behavior is intended to be flexible and provides options to the host application to:

1. opt Github repositories in/out of this feature
1. provide a mechanism to opt individual Pull Requests in/out of the "Review Stack" feature based on Pull Request labels
1. configure how Review Environments are provisioned - at either globally for the host application or on a per-repository basis

Anecdotes
---------

We've been using this feature since December 2019 to serve workload across our entire development staff. Here are some metrics to give a sense of how exercised this feature is within our organization. Our host application:

- deploys review stacks for 10 distinct repositories
- has created over 2_500 Review Stacks and dynamically managed the (de)provisioning of their Review Environments
- has performed over 13_400 deployments of Review Stack commits to their Review Environments

References
----------

- https://devcenter.heroku.com/articles/pipelines#review-apps
- https://github.com/Shopify/shipit-engine/issues/960
- https://github.com/Shopify/shipit-engine/issues/1093
- https://github.com/Shopify/shipit-engine/issues/1076
- https://github.com/Shopify/shipit-engine/pull/967

### CODE REVIEW

File: db/migrate/20200115180018_add_lock_reason_code_to_stacks.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class AddLockReasonCodeToStacks < ActiveRecord::Migration[6.0]
+  def change
+    add_column :stacks, :lock_reason_code, :string
+    add_index :stacks, :lock_reason_code
+  end
+end
```
Comment (benlangfeld) [inline db/migrate/20200115180018_add_lock_reason_code_to_stacks.rb:6]: Can probably drop this file, right?

Comment (benlangfeld) [reply db/migrate/20200115180018_add_lock_reason_code_to_stacks.rb:6]: Indeed, there a few come-and-go migrations like this (where one migration adds a column and another then removes it). We should probably delete these pairs.

File: docs/review_stacks.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    # delete the namespace, etc.
+  end
+
+  def provision?
+    # interrogate status of resources to determin if ample head room exists to provision a new review stack
+  end
+end
+```
+
+In the case where a ``ProvisioningHandler` does not specify a `#provision?` predicate, the provisioning handler will always attempt to provision a Review Stack.
```
Comment (benlangfeld) [inline docs/review_stacks.md:63]:
```suggestion
In the case where a `ProvisioningHandler` does not specify a `#provision?` predicate, the provisioning handler will always attempt to provision a Review Stack.
```

File: docs/review_stacks.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+The host application **MUST** `#register` - whitelist - the custom ProvisioningHandler in the `Shipit::ProvisioningHandler` registry. This will  most likely happen as part of a shipit-engine initialization routine in the host application. For example:
+
+```ruby
+# <path-to-host-application>/config/initializers/shipit.rb
+ActiveSupport::Reloader.to_prepare do
+  ...
+  Shipit::ProvisioningHandler.register(KubernetesProvisioningHandler)
+end
+```
+The custom `KubernetesProvisioningHandler` can the  be used in a project's `shipit.yml`:
```
Comment (benlangfeld) [inline docs/review_stacks.md:74]:
```suggestion
The custom `KubernetesProvisioningHandler` can then be used in a project's `shipit.yml`:
```

File: docs/review_stacks.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+```ruby
+# <path-to-host-application>/config/initializers/shipit.rb
+ActiveSupport::Reloader.to_prepare do
+  ...
+  Shipit::ProvisioningHandler.register(KubernetesProvisioningHandler)
+end
+```
+
+A default, no-op provisioning handler - `Shipit::ProvisioningHandler::Base` - is provided. When the name of an unregistered handler is requested from the `Shipit::ProvisioningHandler` the `Shipit::ProvisioningHandler::UnregisteredProvisiningHander` is used. Effectively, this prevents transitions of the provisioning state machine and locks the Review Stack indicating that an attempt to provision the Review Stack instance was made, but can't be completed until the Review Stack's Provisioning Handler is registered.
```
Comment (benlangfeld) [inline docs/review_stacks.md:99]:
```suggestion
A default, no-op provisioning handler - `Shipit::ProvisioningHandler::Base` - is provided. When the name of an unregistered handler is requested from the `Shipit::ProvisioningHandler` the `Shipit::ProvisioningHandler::UnregisteredProvisioningHandler` is used. Effectively, this prevents transitions of the provisioning state machine and locks the Review Stack indicating that an attempt to provision the Review Stack instance was made, but can't be completed until the Review Stack's Provisioning Handler is registered.
```

File: docs/review_stacks.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+## Default `ProvisioningHandler`s
+
+Register a host-application-wide default provisioning handler - IE how repositories which don't explicitly define a provisoining handler will be provisioned:
+
+```ruby
+# <path-to-host-application>/config/initializers/shipit.rb
+ActiveSupport::Reloader.to_prepare do
+  ...
+  Shipit::ProvisioningHandler.register(KubernetesProvisioningHandler)
```
Comment (benlangfeld) [inline docs/review_stacks.md:95]: Did you mean this?

```suggestion
  Shipit::ProvisioningHandler.default = KubernetesProvisioningHandler
```

File: db/migrate/20200115180018_add_lock_reason_code_to_stacks.rb (hunk 1 above)
Comment (indiebrain) [reply db/migrate/20200115180018_add_lock_reason_code_to_stacks.rb:6]: Was on my list of things to discuss at standup this morning. Thanks for the sanity check. 

> 🍇 🧠 think alike

File: app/serializers/shipit/commit_serializer.rb [hunk 6]
```diff
@@ -7,7 +7,7 @@ class CommitSerializer < ShortCommitSerializer
     has_one :author
     has_one :committer
 
-    attributes :additions, :deletions, :authored_at, :committed_at, :html_url, :pull_request, :status, :deployed
+    attributes :additions, :deletions, :authored_at, :committed_at, :html_url, :merge_request, :status, :deployed
```
Comment (indiebrain) [inline app/serializers/shipit/commit_serializer.rb:10]: I believe these bits need to be removed from the candidate branch as well - reference https://github.com/powerhome/shipit-engine/commit/b94d49cfe3dc3c60b7ee46dc26abcf41732f9b3a

File: db/migrate/20200115180018_add_lock_reason_code_to_stacks.rb (hunk 1 above)
Comment (indiebrain) [reply db/migrate/20200115180018_add_lock_reason_code_to_stacks.rb:6]: https://github.com/powerhome/shipit-engine/pull/72

File: app/serializers/shipit/commit_serializer.rb (hunk 6 above)
Comment (indiebrain) [reply app/serializers/shipit/commit_serializer.rb:10]: https://github.com/powerhome/shipit-engine/pull/73

File: db/migrate/20200714210043_rename_provision_pr_stacks.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class RenameProvisionPrStacks < ActiveRecord::Migration[6.0]
+  def change
+    rename_column :repositories, :provision_pr_stacks, :review_stacks_enabled
```
Comment (benlangfeld) [inline db/migrate/20200714210043_rename_provision_pr_stacks.rb:3]: Do we want to collapse this into the migration which added the column?

**Review by benlangfeld — APPROVED** [review]

File: db/migrate/20200714210043_rename_provision_pr_stacks.rb (hunk 7 above)
Comment (indiebrain) [reply db/migrate/20200714210043_rename_provision_pr_stacks.rb:3]: Damn. I thought I got'm all. Thanks for the 👀

Comment (indiebrain) [reply db/migrate/20200714210043_rename_provision_pr_stacks.rb:3]: https://github.com/powerhome/shipit-engine/pull/74

**Comment by indiebrain** [comment]
Closed since the upstream PR was merged into master here https://github.com/Shopify/shipit-engine/pull/1102

**Pull request closed without merging**